core/src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -163,81 +165,75 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
     dagRequest
   }
 
-  def extractColumnFromFilter(tiFilter: TiExpr, result: ArrayBuffer[TiColumnRef]): Unit =
-    tiFilter match {
-      case fun: TiScalarFunction =>
-        fun.getArgs.foreach(extractColumnFromFilter(_, result))
-      case col: TiColumnRef =>
-        result.add(col)
-      case _ =>
-    }
+  def referencedTiColumns(expression: TiExpr): Seq[TiColumnRef] = expression match {
+    case f: TiScalarFunction => f.getArgs.asScala.flatMap { referencedTiColumns }
+    case ref: TiColumnRef    => Seq(ref)
+    case _                   => Nil
+  }
 
-  def filterToDAGRequest(
+  private def filterToDAGRequest(
     filters: Seq[Expression],
     source: TiDBRelation,
     dagRequest: TiDAGRequest = new TiDAGRequest(pushDownType(), timeZoneOffset())
   ): TiDAGRequest = {
-    val tiFilters: Seq[TiExpr] = filters.collect { case BasicExpression(expr) => expr }
+    val tiFilters = filters.collect { case BasicExpression(expr) => expr }.asJava
     val scanBuilder: ScanBuilder = new ScanBuilder
     val tableScanPlan =
-      scanBuilder.buildTableScan(JavaConversions.seqAsJavaList(tiFilters), source.table)
+      scanBuilder.buildTableScan(tiFilters, source.table)
     val scanPlan = if (allowIndexDoubleRead()) {
       // We need to prepare downgrade information in case of index scan downgrade happens.
-      tableScanPlan.getFilters.foreach(dagRequest.addDowngradeFilter)
-      scanBuilder.buildScan(JavaConversions.seqAsJavaList(tiFilters), source.table)
+      tableScanPlan.getFilters.asScala.foreach { dagRequest.addDowngradeFilter }
+      scanBuilder.buildScan(tiFilters, source.table)
     } else {
       tableScanPlan
     }
 
     dagRequest.addRanges(scanPlan.getKeyRanges)
-    scanPlan.getFilters.foreach(dagRequest.addFilter)
+    scanPlan.getFilters.asScala.foreach { dagRequest.addFilter }
     if (scanPlan.isIndexScan) {
       dagRequest.setIndexInfo(scanPlan.getIndex)
     }
     dagRequest
   }
 
-  def addSortOrder(request: TiDAGRequest, sortOrder: Seq[SortOrder]): Unit =
-    if (sortOrder != null) {
-      sortOrder.foreach(
-        (order: SortOrder) =>
-          request.addOrderByItem(
-            TiByItem.create(
-              BasicExpression.convertToTiExpr(order.child).get,
-              order.direction.sql.equalsIgnoreCase("DESC")
-            )
+  private def addSortOrder(request: TiDAGRequest, sortOrder: Seq[SortOrder]): Unit =
+    sortOrder.foreach { order: SortOrder =>
```

Comment:
why do we delete `null` value check here? If sortOrder is `null`, `foreach` could throw an NPE here.

Reply:
I just saw you change null to nil when you call this function.

Reply:
Yea, please see [this comment thread][1] for more details.

[1]: https://github.com/pingcap/tispark/pull/153#discussion_r158456121

Reply:
Yes. change to nil means empty list. Null value check is unnecessary.